packages/ts/react-crud/src/autogrid.tsx

```diff
@@ -197,7 +197,23 @@ function useColumns(
   });
 
   if (options.customColumns) {
-    columns = [...columns, ...options.customColumns];
+    const columnMap = new Map<string, JSX.Element>();
+    columns.forEach((column) => {
+      const { key } = column;
+      if (key) {
+        columnMap.set(key, column);
+      }
+    });
+    options.customColumns.forEach((customColumn) => {
+      const { key } = customColumn;
```

Comment:
I wonder if we should use `path` instead of `key`. Semantically, `key` is a React thing that is only needed for rendering an array of components. Ideally developers wouldn't have to specify a `key` at all, and it would be set by us. `path` on the other hand is actually about specifying what should be rendered in the column, so there is a semantic relation at least. The `path` values also align with how we specify properties in `visibleColumns`.

Reply:
I agree with this, I also actually had this concern, but I've seen that `key` is already used in existing columns, that's probably why Soroosh chose it. In case we want to use `path`, maybe we should do that everywhere.

Reply:
It sounds like we should narrow the type then. It should not be possible to render `div`s into the grid.

Reply:
Well, assuming there is a type for that. Otherwise we could consider a runtime check.

Reply:
Still, I don't quite understand the original concern. What's wrong with getting the path from the props? It should be accessible using something like `column.props.path`.

Reply:
Well, there's a chance that I don't know how to do it, but when I watch `column.props` there is no `path` set for it, in other word, `column.props.path` undefined at runtime. The closest way to get the same value as path is to read the value of `column.props.value.propertyInfo.name`. 

My concern around both `path` and `column.props.value.propertyInfo.name` is that both can be undefined when creating a computed `GridColumn`. For example the following is working fine without setting any `path`:

```tsx
<AutoGrid
      customColumns={[
           <GridColumn renderer={FullNameRenderer}></GridColumn>,
      ]}
/>
```
What we need is to set "some property" required when defining a custom column so that the developer set it on the `GridColumn` and use it in the `visibleColumns`. Maybe path is a good candidate, but I chose `key` as it is holding the same value (`propertyInfo.name`) for the auto rendered columns, and React checks its uniqueness automatically, and we don't need to worry about accidental overlaps and confusions. But, in general I understand your concern about the `key`. I don't have enough insight about how annoying or abnormal is that to ask developers to set the `key` property for this purpose.

Reply:
Having to set a non-existing path for columns that don't relate to a property is indeed strange. Another argument for using `key` is column groups. In order to deal with the fact that header filters insert a top level column group, you could also pass a column group to custom columns, so the header title would align with the other ones. However column groups don't have the `path` property. For example:
```tsx
    <AutoGrid
      service={PersonListOnlyService}
      model={PersonModel}
      customColumns={[
        <GridColumnGroup key="custom" header="Custom">
          <GridColumn header="" renderer={CustomRenderer} />,
        </GridColumnGroup>
      ]}
    />
```

So I think I would be fine with using `key` for now.